File: tests/serve.py
#!/usr/bin/env python3
"""Threaded static server for the smoke tests.

http.server's default is single-threaded, and the service worker adds a second
wave of requests on top of the page's own. With one thread those queue behind
each other, a fetch times out, and a test fails for a reason that has nothing to
do with the app.
"""

import sys
import time
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs


class Handler(SimpleHTTPRequestHandler):
    # Without an explicit charset the browser guesses, and guesses Latin-1 - every
    # Hebrew string on the page then renders as mojibake.
    extensions_map = {
        **SimpleHTTPRequestHandler.extensions_map,
        '.html': 'text/html; charset=utf-8',
        '.js': 'text/javascript; charset=utf-8',
        '.css': 'text/css; charset=utf-8',
        '.json': 'application/json; charset=utf-8',
        '.webmanifest': 'application/manifest+json; charset=utf-8',
    }

    def do_GET(self):
        # ?slow=N holds the response for N seconds. A site with no signal is the easy
        # case; the one worth testing is a signal that is connected and going nowhere,
        # and only a real slow response exercises the service worker's deadline.
        query = parse_qs(urlparse(self.path).query)
        delay = query.get('slow', [None])[0]
        if delay:
            time.sleep(min(float(delay), 30))
        super().do_GET()

    def end_headers(self):
        # Never let the browser cache during a test run; the service worker is the
        # only caching layer under test.
        self.send_header('Cache-Control', 'no-store')
        super().end_headers()

    def log_message(self, *args):
        pass


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8802
    root = sys.argv[2] if len(sys.argv) > 2 else '.'
    server = ThreadingHTTPServer(('127.0.0.1', port), partial(Handler, directory=root))
    print(f'serving {root} on http://127.0.0.1:{port}', flush=True)
    server.serve_forever()


if __name__ == '__main__':
    main()

File: tests/serve.mjs
// A static file server, in Node, with no dependencies.
//
// The browser suite used to need a second terminal running python3 -m http.server, which
// meant `npm ci && npm run test:smoke` from a fresh clone did nothing but fail. A service
// worker needs a real origin - it does not register over file:// - so the suite cannot
// simply open the files.

import { createServer } from 'node:http';
import { readFile, stat } from 'node:fs/promises';
import { join, extname, normalize } from 'node:path';

const TYPES = {
    '.html': 'text/html; charset=utf-8',
    '.js': 'text/javascript; charset=utf-8',
    '.mjs': 'text/javascript; charset=utf-8',
    '.css': 'text/css; charset=utf-8',
    '.json': 'application/json; charset=utf-8',
    '.webmanifest': 'application/manifest+json; charset=utf-8',
    '.png': 'image/png',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon'
};

// Returns { url, close }. Port 0 lets the OS pick, so two runs at once do not collide
// and a stale server from an earlier run cannot be mistaken for this one.
export function serve(root, port = 0) {
    const server = createServer(async (request, response) => {
        try {
            // normalize() then a prefix check: without it, a request for
            // /../../etc/passwd would be served, and this thing runs on a developer's
            // machine with their own files under it.
            const path = decodeURIComponent(new URL(request.url, 'http://x').pathname);
            let file = normalize(join(root, path === '/' ? '/index.html' : path));
            if (!file.startsWith(normalize(root))) {
                response.writeHead(403).end('no');
                return;
            }

            const info = await stat(file).catch(() => null);
            if (info && info.isDirectory()) file = join(file, 'index.html');

            const body = await readFile(file);
            response.writeHead(200, {
                'Content-Type': TYPES[extname(file)] || 'application/octet-stream',
                // The suite deploys a "new version" mid-run and expects to see it. A
                // cached response would make that test pass or fail on the browser's mood.
                'Cache-Control': 'no-store',
                // The service worker is the point of half these tests, and it will not
                // register without this in some configurations.
                'Service-Worker-Allowed': '/'
            });
            response.end(body);
        } catch (error) {
            response.writeHead(404, { 'Content-Type': 'text/plain' }).end('not found');
        }
    });

    return new Promise(resolve => {
        server.listen(port, '127.0.0.1', () => {
            const { port: actual } = server.address();
            resolve({
                url: `http://127.0.0.1:${actual}`,
                close: () => new Promise(done => server.close(done))
            });
        });
    });
}
